mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/query/SearchQuery.java

```diff
@@ -15,9 +15,19 @@
 public interface SearchQuery<T> {
 
 	/**
-	 * Execute the query and retrieve the results as a {@link List}
+	 * Execute the query and retrieve the results as a {@link SearchResult}.
 	 *
-	 * @return The results as a {@link List} containing one element for each matched entity.
+	 * @return The results.
+	 * @throws org.hibernate.search.util.common.SearchException If something goes wrong while executing the query.
+	 * @throws org.hibernate.HibernateException If something goes wrong while fetching results from the database.
+	 * @throws javax.persistence.PersistenceException If something goes wrong while fetching results from the database.
+	 */
+	SearchResult<T> getResult();
```

Comment:
So it seems we're on the same page, but just a reminder for readers: what we call "result" in Hibernate Search is the whole thing: hits, total hit count, aggregations, facets, ... While a "hit" is only the part of the result that represents a document matching the query.

I agree the names you proposed are more correct. What I was trying to do here is to still more or less fit the JPA naming, so as not to be too confusing.

If we are to give up on that, then yes, I would pick different names. But I would go even further, because I want to make it extra clear that calling one of these methods does not just "get" the hits from a Java object that is already there, fully populated: it actually executes the query.
I would also get rid of the `getSingleHit` variant, since returning an Optional provides the same feature, and more.

So ideally, if we forget about using the same naming as JPA, I would use:

* `SearchResult<T> fetch()`
* `List<T> fetchHits()` (=> `fetch().getHits()`)
* `Optional<T> fetchSingleHit()`  (=> `fetch().getHits()`, but returns an Optional for 0 or 1 hit, throws an exception if more than 1 hit)
* `long fetchCount()` (=> `fetch().getHitCount()`, but optimized to not fetch any hit)
* variants of `fetch()` and `fetchHits` accepting two paramters to handle paging: `fetch(long limit)`, `fetch(long limit, long offset)`... **OR** at least rename `setFirstResult` to `setHitOffset` and `setMaxResults` to `setHitLimit`.

Reply:
TBH, I don't know.

I suppose people would expect an API similar to what they are used to with JPA but it's really 2 completely different things.

Another thing is that a lot of people are using layers on top of JPA so maybe they are not that used to the JPA API.

I like your `fetch(limit, offset)` proposal.

Reply:
I added a few commits to use `fetch(...)` as I proposed.